02_knight_and_princess/robert_gaisin/main.cpp

```diff
@@ -0,0 +1,196 @@
+#include <ctime>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+using VertexId = int;
+using EdgeId = int;
+
+std::ostream& operator<<(std::ostream& out, const vector<int>& int_vector) {
+  for (auto it = int_vector.begin(); it != int_vector.end(); ++it) {
+    out << *it;
+    if (it != int_vector.end() - 1) {
+      out << ", ";
+    }
+  }
+  return out;
+}
+
+struct Vertex {
+  const VertexId id;
+  int depth = 0;
+  vector<EdgeId> edge_ids;
+};
+
+std::ostream& operator<<(std::ostream& out, const Vertex& vertex) {
+  out << "{" << endl;
+  out << "      \"id\": " << vertex.id << "," << endl;
+  out << "      \"edge_ids\": [" << vertex.edge_ids << "]" << endl;
+  out << "    }";
+  return out;
+}
+
+struct Edge {
+  const EdgeId id;
+  VertexId begin = 0;
+  VertexId end = 0;
+};
+
+std::ostream& operator<<(std::ostream& out, const Edge& edge) {
+  out << "{" << endl;
+  out << "      \"id\": " << edge.id << "," << endl;
+  out << "      \"vertex_ids\": [" << edge.begin << ", " << edge.end << "]"
+      << endl;
+  out << "    }";
+  return out;
+}
+
+class Graph {
+ public:
+  void add_vertex();
+
+  void add_edge(VertexId begin, VertexId end);
+
+  void add_edge_in_vertex(VertexId vertex_id, EdgeId edge_id);
+
+  const vector<Vertex>& vertices() const { return vertices_; }
+  const vector<Edge>& edges() const { return edges_; }
+
+ private:
+  VertexId num_of_vrt_ = 0;
+  EdgeId num_of_edg_ = 0;
+
+  VertexId next_vertex_id() { return num_of_vrt_++; }
+  EdgeId next_edge_id() { return num_of_edg_++; }
+
+  vector<Vertex> vertices_;
+  vector<Edge> edges_;
+};
+
+void Graph::add_vertex() {
+  Vertex vertex = {next_vertex_id()};
+  vertices_.push_back(vertex);
```

Comment:
Когда ты добавляешь в коллекцию только что сконструированный объекть, для этого есть специальный метод `emplace`, который, как и конструктор, принимает в себя список аргументов для конструировании нового объекта. Таким образом можно избежать лишней переменной, и написать меньше кода. А меньше кода - всегда лучше.
```suggestion
  vertices_.emplace_back(next_vertex_id());
```